Add asymmetric energy dispersion/migration support

luca-dib opened the pull request:
This adds functions for calculating the asymmetric energy dispersion and migration matricies to `pyirf.irf.energy_dispersion` and creating FITS table HDUs for the asymmetric energy dispersion to `pyirf.io.gadf`.

The functions are separated into functions for a `theta` / `phi` and `lon` / `lat` FoV binning for the direct calculation of energy dispersion and migration and the HDUs:
- `energy_dispersion_asymmetric_polar`
- `energy_dispersion_asymmetric_lonlat`
- `energy_migration_matrix_asymmetric_polar`
- `energy_migration_matrix_asymmetric_lonlat`
- `create_energy_dispersion_asymmetric_polar_hdu`
- `create_energy_dispersion_asymmetric_lonlat_hdu`

Whereas the function for conversion from dispersion to migration `energy_dispersion_to_migration_asymmetric` does not require this.

luca-dib committed:
Add asymmetric edisp functions

luca-dib committed:
Asymmetric edisp, support in normalization func

luca-dib committed:
Add tests for asymmetric edisp

luca-dib committed:
Minor style changes

luca-dib committed:
Add HDU output for asymmetric edisp

luca-dib committed:
Add asymmetric edisp functions

luca-dib committed:
Add missing imports

luca-dib committed:
Minor style changes

luca-dib committed:
Add functions for asymmetric migration matrix

- energy_migration_matrix_polar calculates the migration matrix with FoV binning in
  true_source_fov_offset and true_source_fov_position angle of the FoV
- energy_migration_matrix_lonlat uses true_source_fov_longitude and true_source_fov_latitude

luca-dib committed:
Add functions to __all__

luca-dib committed:
Add unit tests for asymmetric energy migration matrix

luca-dib committed:
Asymmetric dispersion to migration matrix function

luca-dib committed:
Add unit tests for asymmetric disp to migra

luca-dib committed:
Fix typos and style changes

luca-dib committed:
Minor style changes

luca-dib committed:
Add newsfragment

luca-dib committed:
Make newsfragment more detailed

luca-dib committed:
Changed newsfragment to proper PR number

kosack commented on pyirf/io/gadf.py:
```diff
@@ -202,6 +204,130 @@ def create_energy_dispersion_hdu(
+    **header_cards,
+):
+    """
+    Create a fits binary table HDU in GADF format for the asymmetric energy dispersion with polar FoV coordinates.
```
I have a small worry about the terminology: Since it's not obvious that "asymmetric" here means "non-radially-symmetric", i.e. with an azimuthal dependence, it would be good to explain that better.  Otherwise, i would think "asymmetric energy dispersion" is asymmetric in  true energy, which it is, but that is not what we are talking about here.    Maybe just use  similar terminology for describing spatial models as gammapy:  1D vs 2D, i.e.:
* 1D-polar: model is radially symmetric, i.e. no azimuthal dependance, _f(r)_
* 2D-polar: model depends on radius and azimuth, _f(r,phi)_
* 2D-cartesian: projected fov_lon and fov_lat dependance, _f(x,y)._

luca-dib replied:
Thanks for your input!
Yes, I can see there certainly is some ambiguity in just calling it "asymmetric" and I agree that there might be a better way to express this.
Gammapy does not use the "nD" terminology to describe just the spatial axes though (at least for the IRFs), but includes the energy axis as well, e.g. 
- `EnergyDispersion2D` axes: `energy_true` and `offset`, `migra` (This implies to me the `migra` axis is ignored. Is this intentional or an inconsistency?)
- `PSF3D` axes: `energy_true`, `offset` and `rad`.

I definitely think we should be consistent in the naming convention and that there is probably value in also being consistent with gammapy in the future. Whatever way is decided on should be applied to the rest of the IRF package in my opinion, but the details warrant some discussion. For reference, in the non-radially symmetric effective area I also chose the "3D" naming convention, see #281.

luca-dib replied:
> I think the "ND" should refer to the dimensions in the phase space in which an IRF component is stored, excluding intrinsic axes of the IRF component itself.

This also seems reasonable to me, however then the GADF PSF_TABLE breaks with this again, as you said of course.

I would propose that for this PR in accordance with the GADF I change the naming to "3D", however this leaves the PSF for which I would also soon want to create a PR. Is this worth an issue to discuss the naming etc.?

kosack replied:
Yes, I agree with the dimensionality being the spatial+energy, so my comment above becomes just d+1. 

But you are right... for PSF you really need to describe both the internal (source coordinate) and external (field-of-view coordinate) dimensionality to be fully unambiguous.   Not sure the best way to name those... e.g. a model with a "2D" distribution of the PSF that changes with radius and azimuth in the FOV:   _PSF2D_3D_?, _2DPSF3D_?.  It gets ugly!  @bkhelifi let's try to get this right for VODF...

kosack commented on pyirf/io/gadf.py:
```diff
@@ -202,6 +204,130 @@ def create_energy_dispersion_hdu(
+        Bin edges in true energy
+    migration_bins: numpy.ndarray
+        Bin edges for the relative energy migration (``reco_energy / true_energy``)
+    fov_offset_bins: astropy.units.Quantity[angle]
```
Docstring doesn't match signature: cut and paste error?

luca-dib committed:
Change "asymmetric" in function defs to "3d"

luca-dib committed:
Add unit tests for edisp_3d (xfail for now)

luca-dib committed:
Style fixes

luca-dib committed:
Merge branch 'cta-observatory:main' into edisp_asymmetric

maxnoe commented on pyirf/io/gadf.py:
```diff
@@ -202,6 +204,130 @@ def create_energy_dispersion_hdu(
+    header["HDUCLAS4"] = "EDISP_3D"
+    header["DATE"] = Time.now().utc.iso
+    idx = edisp.colnames.index("MATRIX") + 1
+    header[f"CREF{idx}"] = "(ENERG_LO:ENERG_HI,MIGRA_LO:MIGRA_HI,DETX_LO:DETX_HI,DETY_LO:DETY_HI)"
```
I'd put the migra axis last, see https://github.com/open-gamma-ray-astro/gamma-astro-data-formats/issues/178

luca-dib replied:
I see the point, however in that case for consistency's sake the same should be applied to the 2D energy dispersion as well in my opinion. After checking this causes problems with the current tests for IO, specifically with the `test_cref` function. Will try to resolve these.

maxnoe replied:
I agree in principle, but we are in a different situation with the 2d energy dispersion: we have it in GADF with the preferred axis order defined, so we should follow it. That this order is not the one we would prefer is an issue for GADF.

The 3D is new, so we should do what we think is most appropriate.

luca-dib committed:
Address comments

Axis order for 3D EDISP is now changed to put the intrinsic axis last.
Relevant tests updated accordingly.
3D EDISP is no longer wrongly referred to as part of GADF definitions.

luca-dib committed:
Fix missing word in changes file

luca-dib committed:
Merge branch 'cta-observatory:main' into edisp_asymmetric

luca-dib committed:
Add edisp3d HDUs to cref test

luca-dib committed:
Constrain gammapy version to <2.1, where astropy regions <0.12

luca-dib committed:
Lift gammapy constraint, put explicit constraint for regions<0.12

luca-dib committed:
Fix for removed np.cross 2-component vector functionality

luca-dib committed:
Add changelog

luca-dib committed:
Add changelog

kosack committed:
Apply suggestion from @kosack
